util/hot_backup.cc

```diff
@@ -0,0 +1,654 @@
+            //   ScanDB() makes calls ... causing trigger to reset too
+            //   early)
+            gHotBackup->HotBackupScheduled();
+            gHotBackup->HotBackupScheduled();
+
+            // Log() routes to syslog as an error
+            Log(NULL, "leveldb HotBackup triggered.");
+            gPerfCounters->Inc(ePerfBackupStarted);
+
+            DBList()->ScanDBs(true,  &DBImpl::HotBackup);
+            DBList()->ScanDBs(false, &DBImpl::HotBackup);
+
+            // reduce to one instance count now that events posted
+            //  (trigger reset now enabled)
+            gHotBackup->HotBackupFinished();
+        }   // if
+    }   // if
+
+    return;
+
+}   // CheckHotBackupTrigger
+
+
+/**
+ * Look for external trigger
+ */
+bool
+HotBackup::IsTriggerSet()
+{
+    bool ret_flag;
+    int ret_val;
+
+    //
+    // This is code polls for existance of /etc/riak/hot_backup
+    //
+    ret_val=access(GetTriggerPath(), F_OK);
+    ret_flag=(-1!=ret_val);
+
+    return(ret_flag);
+
+}   // HotBackup::IsTriggerSet
+
+
+/**
+ * Let external trigger know that hot backup is complete.  Put
+ *  message to syslog if unable to delete trigger file
+ */
+void
+HotBackup::ResetTrigger()
+{
+    bool ret_flag;
+    int ret_val;
+
+    ret_val=unlink(GetTriggerPath());
+    if (0==ret_val)
+    {
+        // release last counter, enabling future backups
+        HotBackupFinished();
+    }   // if
+    else
+    {
+        // one count left, effectively disabling trigger forever (or next program restart)
+        Log(NULL, "leveldb HotBackup unable to delete trigger file %s", GetTriggerPath());
+        Log(NULL, "leveldb HotBackup now disabled until program restart");
+    }   // if
+
+    return;
+
+}   // ResetHotBackupTrigger
+
+
+/**
+ * A database instance marks its intent to backup
+ */
+void
+HotBackup::HotBackupScheduled()
+{
+    inc_and_fetch(&m_JobsPending);
+
+    return;
+
+}   // HotBackupScheduled
+
+
+/**
+ * A database instance marks that its backup completed
+ */
+void
+HotBackup::HotBackupFinished()
+{
+    int ret_val;
+
+    ret_val=dec_and_fetch(&m_JobsPending);
+
+    // 1 means CheckHotBackupTrigger()'s counter is only
+    //   one left
+    if (1==ret_val)
+    {
+        // hmm, call file sync() at this point?
+        Log(NULL, "leveldb HotBackup complete.");
+        ResetTrigger();
+    }   // if
+
+    return;
+
+}   // HotBackupFinished
+
+
+/**
+ * Routine called by DBList()->ScanDBs.  This has full context of one open
+ *  leveldb database.  Validates then schedules a backup job via compaction
+ *  threads.
+ */
+void
+DBImpl::HotBackup()
+{
+    bool create_backup_event(false);
+
+    // hold mutex while reviewing database state
+    //  (note:  never good to call Log() while holding mutex
+    //   but these are unlikely error cases)
+    {
+        // get mutex,
+        MutexLock l(&mutex_);
+
+        // set backup flag && see if
+        assert(false==hotbackup_pending_);
```

Comment:
Should this assert be removed (or at least commented out)? Didn't we recently have a weird "riak dies every so often" bug caused by an assert that eleveldb fails to disable in a release build?